src/Coordination/SnapshotableHashTable.h

```diff
@@ -3,32 +3,96 @@
 #include <Common/HashTable/HashMap.h>
 #include <Common/ArenaWithFreeLists.h>
 #include <Common/ArenaUtils.h>
-#include <unordered_map>
 #include <list>
-#include <atomic>
 
 
 namespace DB
 {
 
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
 template<typename V>
 struct ListNode
 {
     StringRef key;
     V value;
 
-    /// Monotonically increasing version info for snapshot
-    size_t version{0};
-    bool active_in_map{true};
-    bool free_key{false};
+    /// |*                *            ****** |
+    ///  ^                ^            ^
+    ///  active_in_map    free_key     version
+    ///  (1 byte)         (1 byte)     (6 bytes)
+    uint64_t node_metadata = 0;
+
+    void setInactiveInMap()
+    {
+        node_metadata &= ~active_in_map_mask;
+    }
+
+    void setActiveInMap()
+    {
+        node_metadata |= active_in_map_mask;
+    }
+
+    bool isActiveInMap()
+    {
+        return node_metadata & active_in_map_mask;
+    }
+
+    void setFreeKey()
+    {
+        node_metadata |= free_key_mask;
+    }
+
+    bool getFreeKey()
+    {
+        return node_metadata & free_key_mask;
+    }
+
+    uint64_t getVersion()
+    {
+        return node_metadata & version_mask;
+    }
+
+    void setVersion(uint64_t version)
+    {
+        if (version > version_mask)
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "Snapshot version {} is larger than maximum allowed value {}",
+                version,
+                ListNode<V>::version_mask);
+
+        node_metadata &= ~version_mask;
+        node_metadata |= version;
+    }
+
+    static constexpr uint64_t active_in_map_mask = static_cast<uint64_t>(1) << 63;
+    static constexpr uint64_t free_key_mask      = static_cast<uint64_t>(1) << 62;
+    static constexpr uint64_t version_mask       = ~(static_cast<uint64_t>(3) << 62);
 };
 
 template <class V>
 class SnapshotableHashTable
 {
 private:
+    struct GlobalArena
+    {
+        char * alloc(const size_t size)
```

Comment:
Why not returning `std::unique_ptr<char[]>` ?

Reply:
because I don't want to store that pointer anywhere but inside the `StringRef key`, it will only take extra space
also, this way it's compatible with other Arenas and it's able to use e.g. `copyStringInArena` so I can change them easily if needed